src/oidc.rs

```diff
@@ -73,33 +90,52 @@ impl Into<Token> for TokenResponse {
     }
 }
 
+/// Construct a token exchange request object
+/// For [PKCE flow](https://tools.ietf.org/html/rfc7636#section-4.1) pass in the `code_verifier`
+/// and omit the `client_secret`.
+///
+/// For [authorization code flow](https://auth0.com/docs/flows/authorization-code-flow) pass in
+/// `client_secret` and omit `code_verifier`.
+///
+/// Also supports edge cases where i.e. a development machine requires both
+/// `client_secret` and `code_verifier`. Just pass in them both in such case.
 pub fn exchange_token_request<ReqUri, RedirectUri>(
     uri: ReqUri,
     redirect_uri: RedirectUri,
     client_id: &str,
-    client_secret: &str,
     auth_code: &str,
+    client_secret: Option<&str>,
+    code_verifier: Option<&str>,
 ) -> Result<Request<Vec<u8>>, RequestError>
 where
     ReqUri: TryInto<Uri>,
     RedirectUri: TryInto<Uri>,
 {
-    let body = Serializer::new(String::new())
-        .append_pair("client_id", client_id)
-        .append_pair("client_secret", client_secret)
-        .append_pair("redirect_uri", &into_uri(redirect_uri)?.to_string())
-        .append_pair("grant_type", "authorization_code")
-        .append_pair("code", auth_code)
-        .finish();
+    let mut serializer = Serializer::new(String::new());
+    serializer.append_pair("client_id", client_id);
+    serializer.append_pair("redirect_uri", &into_uri(redirect_uri)?.to_string());
+    serializer.append_pair("grant_type", "authorization_code");
+    serializer.append_pair("code", auth_code);
 
-    let req_body = Vec::from(body);
-    Ok(Request::builder()
-        .method("POST")
-        .uri(into_uri(uri)?)
-        .header(CONTENT_TYPE, "application/x-www-form-urlencoded")
-        .body(req_body)?)
+    if let Some(cs) = client_secret {
+        serializer.append_pair("client_secret", cs);
+    }
+    if let Some(cv) = code_verifier {
+        serializer.append_pair("code_verifier", cv);
+    }
+
+    let body = serializer.finish();
+    http_post_req(body, uri)
 }
 
+/// Construct a token exchange request object
+/// For [PKCE flow](https://tools.ietf.org/html/rfc7636#section-4.1) pass in the `code_verifier`
+/// and omit the `client_secret`.
+///
+/// For [authorization code flow](https://auth0.com/docs/flows/authorization-code-flow) pass in
+/// `client_secret` and omit `code_verifier`.
+///
```

Comment:
I think this doc comment is a copy of the one above for `exchange_token_request` rather than being for `into_uri`.

`into_uri` is internal to the crate so I think we can remove the comment rather than updating it.

Reply:
Yes, also left-over. The function it documented was removed but not the docs. Nicely spotted!